File: src/my_cpp_pkg/src/battery.cpp
#include "rclcpp/rclcpp.hpp"
#include "my_robot_interfaces/srv/set_led.hpp"

using namespace std::chrono_literals;
using namespace std::placeholders;

/**
 * This node simulates a battery level by counting up to 10. 
 * When the counter reaches 4, the battery level is considered empty 
 * and the LED will be turned off.
 */
class BatteryNode : public rclcpp::Node
{
public:
    BatteryNode() : Node("battery")
    {
        set_led_client_ = this->create_client<my_robot_interfaces::srv::SetLed>("set_led");
        battery_level_counter = 0;
        battery_level_empty = false;
        timer_ = this->create_wall_timer(std::chrono::seconds(1), std::bind(&BatteryNode::counter_callback, this));
    }

private:
    /**
     * A timer callback that simulates the battery level by counting up to 10.
     * When the counter reaches 4, the battery level is considered empty and the LED will be turned off. 
     * When the counter reaches 10, the battery level is considered full and the LED will be turned on. 
     * Then the counter is reset to 0 to repeat the cycle.
     */
    void counter_callback()
    {
        battery_level_counter += 1;
        if (battery_level_counter == 4)
        {
            battery_level_empty = true;
            RCLCPP_INFO(this->get_logger(), "Battery level is empty");
            call_set_led(1, true); // Turn off LED when battery is empty
        }
        else if (battery_level_counter == 10)
        {
            battery_level_empty = false;
            RCLCPP_INFO(this->get_logger(), "Battery level is full");
            call_set_led(1, false); // Turn on LED when battery is full
            battery_level_counter = 0;
        }
    }

    /**
     * A service client callback that calls the SetLed service to turn on or off the LED depending on the battery level.
     */
    void call_set_led(int led_number, bool led_state)
    {
        while(!set_led_client_->wait_for_service(1.0s))
        {
            RCLCPP_WARN(this->get_logger(), "Waiting for the server...");
        }
        auto request = std::make_shared<my_robot_interfaces::srv::SetLed::Request>();
        request->led_number = led_number;
        request->led_state = led_state;

        set_led_client_->async_send_request(request);
    }
    
    rclcpp::Client<my_robot_interfaces::srv::SetLed>::SharedPtr set_led_client_;
    rclcpp::TimerBase::SharedPtr timer_;
    int battery_level_counter;
    bool battery_level_empty;
};


int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<BatteryNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/my_py_pkg/my_py_pkg/battery.py
#!/usr/bin/env python3
from functools import partial

import rclpy
from rclpy.node import Node
from my_robot_interfaces.srv import SetLed

"""
A node that acts as a battery to power on an LED depending on
if the battery is fully powered.
"""
class BatteryNode(Node):
    def __init__(self):
        super().__init__("battery")
        self.battery_level_empty = False
        self.led_panel_client = self.create_client(SetLed, "set_led")
        self.timer_ = self.create_timer(0.1, self.battery_timer_callback)
        seconds, nanoseconds = self.get_clock().now().seconds_nanoseconds()
        self.last_time_battery_changed = seconds + nanoseconds / 1000000000.0

    """
    A timer callback that simulates the battery level 
    by using timing methods to considered the battery empty and turn an LED on
    when the timer reaches 4 seconds and the battery full and LED off
    when the timer reaches 6 seconds.
    """
    def battery_timer_callback(self):
        seconds, nanoseconds = self.get_clock().now().seconds_nanoseconds()
        self.time_now = seconds + nanoseconds / 1000000000.0
        if ((self.time_now - self.last_time_battery_changed > 4.0) and not self.battery_level_empty):
            self.battery_level_empty = True
            self.get_logger().info("Battery level is empty")
            self.call_set_led(1, True)
            self.last_time_battery_changed = self.time_now
        elif ((self.time_now - self.last_time_battery_changed > 6.0) and self.battery_level_empty):
            self.battery_level_empty = False
            self.get_logger().info("Battery level is full")
            self.call_set_led(1, False)
            self.last_time_battery_changed = self.time_now


    
    """
    A service client callback that calls the SetLed service to turn on or off the LED depending on the battery level.
    """
    def call_set_led(self, led_number, led_state):
        while not self.led_panel_client.wait_for_service(1.0):
            self.get_logger().warn("Waiting for SetLed server...")

        request = SetLed.Request()
        request.led_number = led_number
        request.led_state = led_state

        self.led_panel_client.call_async(request)


def main(args=None):
    rclpy.init(args=args)
    node = BatteryNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
